src/runtime/config.h

```diff
@@ -0,0 +1,189 @@
+using Override = wcl::optional<T> WakeConfigOverrides::*;
+
+/********************************************************************
+ * Polcies
+ *********************************************************************/
+
+struct VersionPolicy {
+  static constexpr const char* key = "version";
+  static constexpr bool allowed_in_wakeroot = true;
+  static constexpr bool allowed_in_userconfig = false;
+  std::string version = "";
+  static constexpr std::string VersionPolicy::*value = &VersionPolicy::version;
+  static constexpr Override<std::string> override_value = nullptr;
+
+  VersionPolicy() = default;
+  static void set(VersionPolicy& p, const JAST& json);
+};
+
+struct UserConfigPolicy {
+  static constexpr const char* key = "user_config";
+  static constexpr bool allowed_in_wakeroot = true;
+  static constexpr bool allowed_in_userconfig = false;
+  std::string user_config;
+  static constexpr std::string UserConfigPolicy::*value = &UserConfigPolicy::user_config;
+  static constexpr Override<std::string> override_value = nullptr;
+
+  UserConfigPolicy();
+  static void set(UserConfigPolicy& p, const JAST& json);
+};
+
+/********************************************************************
+ * Generic WakeConfig implementation
+ *********************************************************************/
+
+enum class WakeConfigProvenance { Default, Wakeroot, Userconfig, CommandLine };
+
+static inline const char* to_string(WakeConfigProvenance p) {
+  switch (p) {
+    case WakeConfigProvenance::Default:
+      return "Default";
+    case WakeConfigProvenance::Wakeroot:
+      return "Wakeroot";
+    case WakeConfigProvenance::Userconfig:
+      return "Userconfig";
+    case WakeConfigProvenance::CommandLine:
+      return "Commandline";
+  }
+  return "Default";
+}
+
+template <class... Polcies>
+struct WakeConfigImpl : public Polcies... {
+  WakeConfigImpl(const WakeConfigImpl&) = delete;
+  WakeConfigImpl(WakeConfigImpl&&) = delete;
+  WakeConfigImpl() = default;
+
+  std::map<std::string, WakeConfigProvenance> provinence;
+
+ private:
+  struct void_t {};
+
+  template <class... Args>
+  static void call_all(Args... x) {}
+
+  template <class P>
+  void_t emit_each(std::ostream& os) const {
+    auto iter = provinence.find(P::key);
+    auto p = WakeConfigProvenance::Default;
+    if (iter != provinence.end()) {
+      p = iter->second;
+    }
+    os << "  " << P::key << " = '" << this->*P::value << "' (" << to_string(p) << ")" << std::endl;
+    return {};
+  }
+
+  template <class P>
+  static void_t add_wakeroot_key(std::set<std::string>& out) {
+    if (P::allowed_in_wakeroot) {
+      out.emplace(P::key);
+    }
+    return {};
+  }
+
+  template <class P>
+  static void_t add_userconfig_key(std::set<std::string>& out) {
+    if (P::allowed_in_userconfig) {
+      out.emplace(P::key);
+    }
+    return {};
+  }
+
+  template <WakeConfigProvenance p, class P>
+  void_t set_policy(const JAST& json) {
+    // The current semantics want us to
```

Comment:
This comment is unclear.  Needs to be more detailed or removed

Reply:
Yeah it was half written lol